### ISSUE TITLE
Support proxy network

### ISSUE THREAD

**Issue by hanguokai** [google/clasp#8]
To use proxy network, like http(s) or socks5, I use `http_proxy` and `https_proxy` environment variable that are supported by many tools. Currently, clasp can't work in proxy environment, for example `clasp login` and `clasp clone`.

**Comment by grant** [comment]
I'm not sure what this would change would entail.
Can you be more specific about repro steps, some links, and a more detailed proposal on what commands this would be needed for?

It looks like this request would want to add something like:
`--proxy "socks5://127.0.0.1/"`
to every command.

Similar issue: https://github.com/rg3/youtube-dl/issues/402#issuecomment-213744197

**Comment by hanguokai** [comment]
Yes, clasp need to add `--proxy` option, like youtube-dl, curl, gcloud, npm, git, etc. 

`http_proxy` environment variable is the same meaning as `--proxy`, but detect by command if no --proxy option specified.

I am using socks5 protocol, there are others like http, https.

**Comment by grant** [comment]
@hanguokai Can you add proxy rule to the URL `googleapis.com` so that every `googleapis.com` request is proxied?

**Comment by grant** [comment]
Pull requests are welcome.
I'm closing this since I don't see much demand for this feature. Will reopen if there's demand or more discussion.

**Comment by HeavensBlade** [comment]
I would also like to see a proxy option.
Since I'm developing behind a corporate proxy I always need to activate my smartphone hotspot.

**Comment by nepsilon** [comment]
I'm using clasp from a censored internet where all things Google are blocked and need the proxy feature. Using proxy environment variables doesn't work.

**Comment by nepsilon** [comment]
@grant Yes, all blocked domains and IP requires a VPN or proxy. It's very standard in places like China.

**Comment by grant** [comment]
Unfortunately G Suite and many of Google's services are blocked in China. I can't solve that issue. I'm guessing the Firebase tool would probably prefer to not be blocked too to solve that linked issue.

I don't understand why environment variables don't work, do you really need every single tool you use to implement a custom `--proxy` option?

Underneath it all, we use [`googleapis`](https://www.npmjs.com/package/googleapis). It supports a [proxy](https://www.npmjs.com/package/googleapis#using-a-proxy) (`HTTP_PROXY`, `HTTPS_PROXY`). Does that not work?

To implement this feature, I'd need:
- A reasonable number of people to request and use the proxy feature (👍 the issue)
- A design doc for this feature.
- Ideally a pull request with an initial implementation.

To be honest, I have no use for this feature and still don't understand why you can't use a proxy elsewhere in the tech stack.

**Comment by valen22br** [comment]
Hello, I'm running through the same problem. I tried setting up my user variables with proxy settings, end I'm still receiving the error below:
Error retrieving access token: Error: connect ETIMEDOUT 172.217.18.170:443

Is there a workaround for this issue?

**Comment by grant** [comment]
I'm in favor of adding a proxy option, but to be honest, I don't know how to run a proxy and I don't know why env variables don't work like in the `googleapis` README:
```
export HTTP_PROXY=http://localhost:3128
export http_proxy=http://localhost:3128
export https_proxy=http://localhost:3128
export HTTPS_PROXY=http://localhost:3128
```

I did a bit more research.

- If someone wants to try this, add a `proxy` variable to this line of code: https://github.com/google/clasp/blob/master/src/auth.ts#L40
Then rebuild clasp and test it out, and it works, that'd be awesome. Then I can update clasp to include this proxy param.
- If someone (@hanguokai, @nepsilon) describes exact steps on how to download a proxy, see the error in clasp using some tool, I'll try to reproduce the problem and create a solution.
- If you need help with a pull request, or building clasp, LMK too.

This is the most upvoted issue, and it looks like user variables doesn't work, so I'm in favor of support this feature in `clasp`, but I'll need some help for this one.

**Comment by valen22br** [comment]
Thanks for the reply!
I'll take a look at it. I work for a big company that enforces us to use a
proxy. That's why for us, having a way to support proxy is extremely
important. I don't need a specific solution as the --proxy option, but a
watever solution that works behind a proxy.

Best,
Luis
Gustavo

**Comment by valen22br** [comment]
Can you give me directions of how to build clasp locally for testing
purpose? Them if works, you can help me with the pull request.
Best,
Luis Gustavo

**Comment by grant** [comment]
@valen22br Follow these instructions https://github.com/google/clasp#develop
Let me know if you can't build clasp.

**Comment by valen22br** [comment]
@grant the instructions worked for me. I will start playing with the proxy settings and will let you know the results.

**Comment by hanguokai** [comment]
I did some research about google-api-nodejs-client. It says [it support proxy](https://github.com/google/google-api-nodejs-client/tree/master#using-a-proxy).

But in fact, googleapis use [axios](https://github.com/axios/axios) as it's http client library that **doesn't support socks proxies**. So either `export https_proxy=socks5://127.0.0.1:1080` or set socks proxy option in code doesn't work.

In theory, googleapis should support http/https proxy like `export http_proxy=http://localhost:3128`. I didn't research whether there is a problem with this method.

PS: I know `gcloud` command support proxy including socks proxy. And `firebase` command had problem for supporting proxy, but recently it seems support http proxy(not socks proxy) in 4.1.2.

**Comment by grant** [comment]
I don't know the difference between socks proxies and non-socks proxies. Will a non-socks proxy work?

**Comment by valen22br** [comment]
I don't know either. 
From my tests I got the following results:
1. No HTTP_PROXY or HTTPS_PROXY environment setup. In this scenario, the error was: Error retrieving access token: Error: connect ETIMEDOUT 172.217.18.170:443
2. With HTTP_PROXY or HTTPS_PROXY environment setup. Now I got another error: Error retrieving access token: Error: write EPROTO 101057795:error:140770FC:SSL
routines:SSL23_GET_SERVER_HELLO:unknown protocol:openssl\ssl\s23_clnt.c:827:

It seems that we are going from a proxy error to some OpenSSL incompatibility. Do you have such OpenSSL error message before?

**Comment by valen22br** [comment]
The call to the method **client.getToken(authCode)** at auth.ts was not able to provide a return. Do you know how to debug that method?

**Comment by hanguokai** [comment]
> I don't know why env variables don't work like in the googleapis README - grant
> Will a non-socks proxy work? - grant

I wrote a sample to test env variables for googleapis directly. I found googleapis itself has issue for supporting http(non-socks) proxy. Someone filed an issue [here](https://github.com/google/google-api-nodejs-client/issues/998). Axios, googleapis use it as http client library, also has [issue](https://github.com/axios/axios/pull/959) about http proxy.

So when googleapis fixed http proxy issue, let's review this issue again in clasp. Currently, I use VPN as a workaround method.

**Comment by valen22br** [comment]
@hanguokai , good point in your research. My tests also revealed that apparently, the environment variables with proxy settings are working. The problem is when the clasp tries to use the token got from the authentication process to use it against the google API. In this step, it seems that someone, (Google API as you pointed) is messing around with https requests. 
Here in our company, we have a single proxy that can handle both HTTP and https requests without distinction.

I'll take a look at the links you provided. Can you explain your workaround using the VPN behind the corporate proxy?

**Comment by valen22br** [comment]
Hi @grant , 
I was recreating my development environment and now I can't build clasp anymore using the command npm run build or npm run build-fresh. 

Following is the error:
> tsc --project tsconfig.json && npm i -g --loglevel=error

src/auth.ts:1:13 - error TS2300: Duplicate identifier 'fs'.

1 import * as fs from 'fs';
              ~~


src/auth.ts:2:13 - error TS2300: Duplicate identifier 'fs'.

2 import * as fs from 'fs';
              ~~


src/auth.ts:11:10 - error TS2300: Duplicate identifier 'Discovery'.

11 import { Discovery } from 'googleapis/build/src/apis/discovery/v1';
            ~~~~~~~~~


src/auth.ts:12:10 - error TS2300: Duplicate identifier 'Discovery'.

12 import { Discovery } from 'googleapis/build/src/apis/discovery/v1';
            ~~~~~~~~~


npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! @google/clasp@1.5.3 build: `tsc --project tsconfig.json && npm i -g --l
oglevel=error`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the @google/clasp@1.5.3 build script.
npm ERR! This is probably not a problem with npm. There is likely additional log
ging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\GL000017\AppData\Roaming\npm-cache\_logs\2018-08-23T18_37_
47_980Z-debug.log

**Comment by valen22br** [comment]
I'm looking at your code, specifically at auth.ts file, and there are some duplicated lines on it. 
I removed the duplicated lines, and it is building successfully. Why were those lines duplicated?

**Comment by grant** [comment]
@valen22br Ah, that looks like a separate issue (from a separate change). Can we discuss this in a separate GitHub Issue?

**Comment by valen22br** [comment]
Yes, no problem!

**Comment by grant** [comment]
(I moved some off-topic comments to a separate thread to keep this about a proxy FR. See https://github.com/google/clasp/issues/297#issuecomment-415641264)

**Comment by aakasaka** [comment]
It seems googleapis [fixed the problem](https://github.com/googleapis/google-api-nodejs-client/issues/998) with proxy.  
Could you please update the packages (googleapis and google-auth-library) to latest version?

**Comment by aakasaka** [comment]
I was able to `clasp login` and `clasp create` with the latest version, thank you !  
FYI, I still should replace the code `inOnline()` with `true` coz `is-online` module doesn't work with proxy. (mentioned in #336 )

**Comment by grant** [comment]
OK, thanks. I'll remove the `isOnline` feature since it's not super useful and blocks proxy users.

**Comment by valen22br** [comment]
Tomorrow I will run some tests with a proxy configuration.
Thanks for the fix!

**Comment by grant** [comment]
@aakasaka I removed the check to `isOnline` if `HTTP_PROXY/HTTPS_PROXY` is present with `1.6.3`.
Let me know if a proxy works out of the box now, so I can close this issue.

**Comment by hanguokai** [comment]
I simply tested clasp 1.6.3 with login, logout, status, clone, pull and push commands. It seems Ok now for `http_proxy/https_proxy`. Thanks for updating!

**Comment by zoudino** [comment]
Hi, @grant , could you provide a quick step by step tutorial on how to set up the proxy and use the clasp? I have been trying hours on figuring out this issue even after reading this,long discussion.  The error I got is still showing "Error retrieving access token: FetchError: request to https://oauth2.googleapis.com/token failed, reason: connect ETIMEDOUT xxx:443"

**Comment by thedrs** [comment]
I solved this issue, as i had another problem here related to proxy - here is the story:

I am on windows in a very restricted corporate enterprise which has to use an http and https proxy.
We do have normal google services accessible in the browser.
Obviously 'clasp login' does not work without a proxy, it gives:
Error retrieving access token: FetchError: request to https://oauth2.googleapis.com/token failed, reason: connect ETIMEDOUT 216.58.212.202:443
(It does work if i disconnect my computer from corporate and connect to public internet)

So i defined these two env variables in cmd (you can also do it in git bash console, just replace "set" with "export"):
set HTTP_PROXY=[http://proxy.mycompany.com:80](http://proxy.mycompany.com/)
set HTTPS_PROXY=https://proxy.mycompany.com:80/
(our corporate proxy uses port 80 for both http and https !)

this is the error i got:
Error retrieving access token: FetchError: request to https://oauth2.googleapis.com/token failed, reason: write EPROTO EC110000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:c:\ws\deps\openssl\openssl\ssl\record\ssl3_record.c:355:

The solution to my problem was that our https proxy used an http address, so i fixed the set to the following :
set HTTPS_PROXY=http://proxy.mycompany.com:80](http://proxy.mycompany.com/)
(note the http:// and not https:// on the right hand side)

Hope i helped others here too
